fncas/linux_native_jit/linux_native_jit.h

```diff
@@ -0,0 +1,352 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2019 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+// NOTE(dkorolev): This is really an alpha version. "It works on my machines", and then I don't care for now. -- D.K.
+// TODO(dkorolev): Indirect calls to external functions (to avoid dealing with `@plt`) may be suboptimal.
+// TODO(dkorolev): Offsets to offsets, to make sure the instructions have the same opcode length, may be suboptimal.
+// TODO(dkorolev): Look into endianness.
+
+#ifndef FNCAS_LINUX_NATIVE_JIT_LINUX_NATIVE_JIT_H
+#define FNCAS_LINUX_NATIVE_JIT_LINUX_NATIVE_JIT_H
+
+#if defined(__linux__) && defined(__x86_64__)
+
+#define FNCAS_LINUX_NATIVE_JIT_ENABLED
+
+#include <cmath>
+#include <cstdint>
+#include <cstring>
+#include <iostream>
+#include <vector>
+
+#include <sys/mman.h>
+
+#ifdef NDEBUG
+#include <cassert>
+#define LINUX_JIT_ASSERT(x) assert(x)
+#else
+#define LINUX_JIT_ASSERT(x)
+#endif
+
+static_assert(sizeof(double) == 8, "FnCAS Linux native JIT compiler requires `double` to be 8 bytes.");
+
+namespace current {
+namespace fncas {
+namespace linux_native_jit {
+
+// The signature is:
+// * Returns a `double`:              The value of the function.
+// * Uses the `double const* x`:      Input, the parameters vector `x[i]`.
+// * Uses the `double* o`:            Output, first the gradient, then the temporary memory buffer.
+// * Uses the `double (*f[])(double): External functions (`sin`, `exp`, etc.) to be called, to avoid dealing with PLT.
+typedef double (*pf_t)(double const* x, double* o, double (*f[])(double));
+
+constexpr static size_t const kLinuxNativeJITExecutableMageSize = 4096;
```

Comment:
Did you mean `PageSize`?

Reply:
Yeah, I guess so. Thanks! https://github.com/C5T/Current/pull/840/commits/e0d2f4502367228d8b8cfb6a69437544b8c5a1c7